Course manage

Thommond opened the pull request:
Here is the course-manage branch which includes the courseManagement, course edit, and course create pages for teachers. 

On test_courseEditor.py is my test for the feature which is 80% coverage.

CourseEditor.py which is this feature blueprint has all the routes.

Thommond committed:
I started the tests for the courseManagement page.

I added two tests, the edit and the create test.

I then moved the test_app.py to test_portal.py.

Thommond committed:
I completed 4 main tasks.

I updated the portal.py to have the routes needed for the teacher edit page.

I updated the schema.sql with three tables I will most likely need for the
user story including the major, session, and courses.

I added the portal blueprint to the __init__.py.

I finally added some more changes to the tests including creating an
info test for the info route.

Thommond committed:
I completed two main tasks.

I first exited a little on the portal.py and courseMane.html like syntax.

I then made the tables that I need for this user story like sessions,
major, and courses and updated users table in the schema.sql file..

I then made mock data for the tables created in the data.sql file.

Thommond committed:
I completed two main tasks.

I added more info on each route on portal.py.

I then created some needed templates for edit and info routes.

I also polished the data in data.sql and schema.sql.

Thommond committed:
I added some foriegn keys to schema.sql .

I created a courseCreate template.

I editing routeing in portal.py.

I updated the LICENSE with our proper names.

I added more to each template for functionality.

Thommond committed:
I completed a few tasks.

I updated the naming convention in __init__.py

I completed the functionallity of the create and edit functions
for course.

I tweaked the three template pages to match the
functionallity of courseEditor.py

Lastly, I updated data.sql to needed parameters.

Thommond committed:
Merge branch 'master' into course-manage

Thommond committed:
I made edited the templates to match the database.

I created the teachers required function to use
in the courseEditor route in auth.py

I updated courseEditor.py with teacher required.

I edited data.sql for testing.

I added a few styles.

Thommond committed:
Merge branch 'master' into course-manage

Thommond committed:
I added two csv files for sessions and courses.

I fixed the structure and schema.sql .

I lastly added more detail to tests.

Thommond committed:
I finished the tests for the courseEdit, courseCreate, and
courseManage pages. I also created log in and log out functions
to make it easier for test access.

BrendanHorst commented on portal/courseEditor.py:
```diff
@@ -0,0 +1,141 @@
+def course_create():
+    """Allows the teacher to create a course
+    and fill out specifics on course"""
+    Allmajors = get_majors()
```
`Allmajors` should probably be written as `all_majors`

BrendanHorst commented on portal/courseEditor.py:
```diff
@@ -0,0 +1,141 @@
+                courseTitle = request.form['courseTitle']
+                courseDescription = request.form['description']
+                courseCredit = request.form['courseCredits']
+                courseMajor = request.form['major_name']
```
again, not super important, but camel case is JavaScript convention, Python convention is snake case.
That goes for the name of the file, `courseEditor`, too

BrendanHorst commented on portal/courseEditor.py:
```diff
@@ -0,0 +1,141 @@
+                    error = 'You are missing a required field'
+                if not courseDescription:
+                    error = 'You are missing a required field'
+                if not courseMajor:
```
Not necessary, but it may be good to ensure that the major id exists / is an integer before inserting the data.  Since it is a select and not a next input, its not necessary for functionality, but it could be good for security purposes

BrendanHorst commented on portal/courseEditor.py:
```diff
@@ -0,0 +1,141 @@
+                    """,
+                                (title, desc, credit, id)
+                                )
+                    con.commit()
```
It's probably best to only put this inside of the `with` and not everything else.  Since it is a `with` it may not matter, but its best to do as little as possible with a cursor open just in case the code breaks

BrendanHorst commented on portal/db.py:
```diff
@@ -105,13 +110,15 @@ def insert_users_command():
 def mock_db():
     """Seed the database with mock data."""
 
-    data_filepath = os.path.join(os.path.dirname(__file__), os.pardir, "tests", "data.sql")
+    data_filepath = os.path.join(os.path.dirname(
+        __file__), os.pardir, "tests", "data.sql")
```
why did this change to two lines? Not that it matters, just curious why that was explicitly changed, unless it was text editor shenanigans

Thommond replied:
Text editor beautify. I must have not even noticed the change.

BrendanHorst commented on portal/templates/layouts/base.html:
```diff
@@ -9,7 +9,7 @@
     <link rel="stylesheet" href="https://necolas.github.io/normalize.css/latest/normalize.css" type="text/css" />
     <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Crimson+Text|Lato:400,400i,700" />
-    <link rel="stylesheet" href="static/styles.css" type="text/css" />
+    <link rel="stylesheet" href="http://localhost:5000/static/styles.css" type="text/css" />
```
this probably should not be pointing to a local host. `/static/styles.css` or `{{ url_for('static', filename='style.css') }}` would be preferable

BrendanHorst commented on portal/templates/layouts/courseEdit.html:
```diff
@@ -0,0 +1,70 @@
+{% extends 'layouts/base.html' %}
```
this whole file could use proper indentation

BrendanHorst commented on portal/templates/layouts/courseEdit.html:
```diff
@@ -0,0 +1,70 @@
+  please fill out all the fields and save changes.</p>
+
+<ul>
+
+
+  <label for="editTitle">Course Title</label><!--Bold in styles-->
+
+<li> {{ course['course_title'] }}
+
+
+  <input type="text" name="editTitle" required>
+
+
+
+
+</li>
+
+<label for="editDesc">Course Description</label>
+
+<li>{{ course['description'] }}
+
+<input type="text" name="editDesc" required>
+
+
+
+</li>
+
+  <label for="editCredit">Course Credits</label>
+
+<li>{{ course['credits'] }}
+
+  <input type="text" name="editCredit" required>
+
+
+
+</li>
+
+
+
+</ul>
```
the `<label>`s should be inside of the `<li>`s as a label and its corresponding input should be one item on the list
Right now, the unordered list goes:
- label
- list item
- label
- list item
- label
- list item

BrendanHorst commented on portal/templates/layouts/courseInfo.html:
```diff
@@ -0,0 +1,26 @@
+  <header>
+    <h1>{{ course['course_title']}}</h1>
+    <a href="/courseManagement">Back</a>
+    <!--<a class='action' href="{{ url_for(courseInfo.update), id=course['course_num']}}">Edit</a>-->
```
why is this line commented it out? should it be deleted or inserted?

Thommond replied:
courseInfo.html should not exist. It is not part of the user story.

BrendanHorst commented on portal/templates/layouts/courseInfo.html:
```diff
@@ -0,0 +1,26 @@
+
+  <header>
+    <h1>{{ course['course_title']}}</h1>
+    <a href="/courseManagement">Back</a>
```
`url_for()` would be preferable here

Thommond replied:
Okay

BrendanHorst commented on portal/templates/layouts/courseMan.html:
```diff
@@ -0,0 +1,52 @@
+
+<header>
+  <h2>Course Management</h2>
+  <a href="{{ url_for('index')}}">Home</a>
```
this isn't wrong, but it would probably be better to put this on a nav bar on the base template or something, so that you have a home button on every page.  But we haven't really discussed what to put on, or where to put a nav bar yet, so don't worry about it

BrendanHorst commented on portal/templates/layouts/courseCreate.html:
```diff
@@ -0,0 +1,43 @@
+{% extends 'layouts/base.html' %}
```
this file could use some proper indentation

BrendanHorst reviewed (changes requested):
functionality-wise it looks pretty solid.  Just watch out for semantics and formatting: 
JavaScript uses camelCase, while Python should be in snake_case for variables.  I didn't bother marking every single instance of camel casing in python files or capitalized variables, so just double check all your variable names.
Also, html may not break like python, but please indent it.

BrendanHorst commented:
double check that a teacher can't update/delete a course that isn't theirs, even if they can't click on it on the interface

Thommond committed:
I updated nick picky variable names and semantics.

Thommond committed:
Merge branch 'master' into course-manage

Thommond committed:
I merge master and updated some csv convention.

Thommond committed:
Fixed the style sheet link.

Thommond committed:
I added an additional if statement in course_editor.py to ensure
that teachers cannot edit other teachers course.

I updated courses.csv to test the course.

I also wrote a test for the added if statement.

BrendanHorst reviewed (approved):
HTML could still use indentation, but other than that it looks good
